### PR TITLE
Fix #48; separation of training and validation

### LINKED ISSUE #48: Separate code for training from validation and test

### PR DESCRIPTION
**Pull request by tilmankamp**

### CODE REVIEW

**Comment by tilmankamp** [comment]
@lissyx @kdavis-mozilla : WIP for the TensorBoard milestone.

**Comment by kdavis-mozilla** [comment]
@tilmankamp I'll review this along with Ruben's pull request #46 on Monday, then integrate the changes from my pull request #50 and Ruben's pull request #46 into this pull request and ask you for a review of the code, then once you give your OK, we can do one big merge. (I don't think waiting on Andre for the review of pull request #50 is time efficient at this point.)

File: DeepSpeech.ipynb [hunk 1]
```diff
@@ -1105,8 +1306,123 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 34,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "written = False\n",
+    "logs_dir = \"logs\"\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: Why define `logs_dir`twice. Once here and once for the code above...
```
log_dir = '%s/%s' % ("logs", time.strftime("%Y%m%d-%H%M%S"))
```
Seems like you should define 
```
logs_dir = "logs"
```
once then use it in both places.

File: DeepSpeech.ipynb [hunk 2]
```diff
@@ -1021,80 +1170,132 @@
...
+    "    for epoch in range(training_iters):\n",
+    "        # Define total accuracy for the epoch\n",
+    "        total_accuracy = 0\n",
     "        \n",
-    "        # Create session in which to execute\n",
-    "        session = tf.Session(config=tf.ConfigProto(allow_soft_placement=True, log_device_placement=True))\n",
+    "        # Validation step to determine the best point in time to stop\n",
+    "        if epoch % validation_step == 0:\n",
+    "            _, last_validation_wer = print_wer_report(session, \"Validation\", validation_tower_decodings, validation_tower_labels)\n",
+    "            # TODO: Determine on base of WER, if model starts overfitting\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: > Determine on base of WER, if model starts overfitting

I don't think this makes sense. 

I read this as saying: If the model starts overfitting, then stop training only once the validation WER is below the target threshold.

Overfitting can occur and the validation WER could be above the threshold set and it can stay there. So, this could lead to an infinite loop.

We don't have to do this now, but we could use early stopping using the ValidationMonitor[[1]](https://www.tensorflow.org/versions/r0.10/tutorials/monitors/index.html#early-stopping-with-validationmonitor) and base early stopping on WER.

File: DeepSpeech.ipynb [hunk 3]
```diff
@@ -142,7 +144,7 @@
    },
    "outputs": [],
    "source": [
-    "dropout_rate = 0.05  # TODO: Validate this is a reasonable value"
+    "dropout_rate = 0.01  # TODO: Validate this is a reasonable value"
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: Done

File: DeepSpeech.ipynb [hunk 4]
```diff
@@ -1105,7 +1264,8 @@
    },
    "outputs": [],
    "source": [
-    "train()"
+    "# Test network\n",
+    "print_wer_report(session, \"Test\", forward(session, data_sets.test))"
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: This was already done in a later commit - sorry for the noise.

File: DeepSpeech.ipynb [hunk 5]
```diff
@@ -1105,8 +1306,123 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "with open('%s/%s' % (log_dir, 'hyper.json'), 'w') as dump_file:\n",
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: Already had a similar doubt about this. However: The only way I see is to have all the parameters already defined within a dictionary. But this would hinder us from defining them in a distributed way across the document.

File: DeepSpeech.ipynb (hunk 2 above)
Comment (tilmankamp) [inline DeepSpeech.ipynb]: I think we are/were on the same page here. Just removed the comment.

File: DeepSpeech.ipynb (hunk 1 above)
Comment (tilmankamp) [inline DeepSpeech.ipynb]: There is a ```log``` and a ```logs``` one. I missed to introduce and reuse the ```logs_dir``` already in the upper cell for the helpers.

File: DeepSpeech.ipynb [hunk 6]
```diff
@@ -1003,6 +1032,126 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "def forward(session, data_set):\n",
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: Good find (session)! Also renamed it.

File: DeepSpeech.ipynb [hunk 7]
```diff
@@ -1105,8 +1306,123 @@
...
+    "written = False\n",
+    "logs_dir = \"logs\"\n",
+    "\n",
+    "# All direct sub directories of the logs directory\n",
+    "dirs = [os.path.join(logs_dir, o) for o in os.listdir(logs_dir) if os.path.isdir(os.path.join(logs_dir, o))]\n",
+    "\n",
+    "# Let's first populate a temporal file and rename it afterwards - guarantees an interruption free web experience\n",
+    "nhf = '%s/%s' % (logs_dir, 'new_hyper.js')\n",
+    "\n",
+    "with open(nhf, 'w') as dump_file:\n",
```
Comment (tilmankamp) [inline DeepSpeech.ipynb]: Done

**Comment by tilmankamp** [comment]
@kdavis-mozilla : See above for some detail regarding the code. Fixed also all markup.

File: DeepSpeech.ipynb [hunk 8]
```diff
@@ -1021,80 +1177,158 @@
...
-    "            \n",
-    "            # Checkpoint the model\n",
-    "            if (epoch % checkpoint_step == 0) or (epoch == training_iters - 1):\n",
-    "                checkpoint_path = os.path.join(checkpoint_dir, 'model.ckpt')\n",
-    "                print \"Checkpointing in directory\", \"%s\" % checkpoint_dir\n",
-    "                saver.save(session, checkpoint_path, global_step=epoch)\n",
+    "        # Print progress message\n",
+    "        if epoch % display_step == 0:\n",
+    "            print \"Epoch:\", '%04d' % (epoch+1), \"avg_cer=\", \"{:.9f}\".format((total_accuracy / total_batch))\n",
+    "            _, last_train_wer = print_wer_report(session, \"Training\", tower_decodings, tower_labels)\n",
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: According to the example of [[1](https://www.tensorflow.org/versions/r0.9/tutorials/mnist/pros/index.html#train-and-evaluate-the-model)], this will produce incorrect results as it is printing results from the training data where `dropout_rate=0.05` and not `dropout_rate=0.0`.

This is due to the fact that the line
```
tower_decodings, tower_labels, tower_gradients, tower_loss, accuracy =
    get_tower_results(n_steps, data_sets.train, optimizer)
```
used to create `tower_decodings`, `tower_labels`, `tower_gradients`, `tower_loss`, and `accuracy` passes an `optimizer` in which causes `dropout_rate=0.05` to be used. 

It seems like it might make more sense to introduce something like..
```
train_tower_decodings, train_tower_labels = decode_batch(data_sets.train)
```
then print the results from this like you do for the validation data.

Comment (tilmankamp) [inline DeepSpeech.ipynb]: Another good catch! However: I took a different approach by re-introducing a feed_dict. I think it turned out to be quite elegant. Please have a look.

**Comment by kdavis-mozilla** [comment]
@tilmankamp I'll take a look.

File: DeepSpeech.ipynb [hunk 9]
```diff
@@ -148,7 +148,16 @@
...
+    "dropout_rate = 0.05  # TODO: Validate this is a reasonable value\n",
+    "\n",
+    "# This global placeholder will be used for all dropout definitions\n",
+    "dropout_rate_placeholder = tf.placeholder(tf.float32)\n",
+    "\n",
+    "# The feed_dict used for training employs the given dropout_rate\n",
+    "feed_dict_train = { dropout_rate_placeholder: dropout_rate }\n",
+    "\n",
+    "# While the feed_dict used for validation, test and train progress reporting employs zero dropout\n",
+    "feed_dict = { dropout_rate_placeholder: 0.0 }"
```
Comment (kdavis-mozilla) [inline DeepSpeech.ipynb]: This is a much cleaner way of doing things, and I think for small data sets, like the single scalar `dropout_rate`, it's likely fine.

However, one reason I moved away from `feed_dict` when going to the multi-gpu code is that  `feed_dict` is known to be slow[[1](https://github.com/tensorflow/tensorflow/issues/2919)]. In the thread [nanddalal](https://github.com/nanddalal) found using `feed_dict` with a GPU was about 4-5 times as slow as not using it on a GPU.

Quoting [yaroslavvb](https://github.com/yaroslavvb) (who was part of Google Brain and is part of OpenAI) from the thread referenced above:

> Feed_dict does a single-threaded memcpy of contents from Python runtime into TensorFlow runtime. If data is needed on GPU, then you'll have an additional CPU->GPU transfer. I'm used to seeing up to 10x improvement in performance when switching from feed_dict to native TensorFlow (Variable/Queue)

If we use `feed_dict`, we need to be careful not to put more stuff into `feed_dict`. (Which I guess is easiest if it's not there :-) )

You can make the call on keeping `feed_dict` or not. But just tell me your decision so I can start integrating my pull request in to yours before we merge yours.

Comment (tilmankamp) [inline DeepSpeech.ipynb]: As we talk about a single scalar value, **I vote for keeping it** - at least for now. If this is really getting a problem, it could be fixed easily.
